### PR TITLE
Add functionality to promote nested dependencies to top level.

### PR DESCRIPTION
**Pull request by jpakkane**

### CODE REVIEW

**Comment by nirbheek** [comment]
I was thinking about this, and this is good to have. However, I think we should have a good default story for subproject wraps too. It could be something as simple as:

1. For each dependency that needs a fallback, look for wraps in `subprojects/` (as usual)
1. For each subproject which needs a new fallback dependency, look for a wrap in the source tree of that subproject, execute, and cache it (new behaviour)
1. If the same dependency is needed again in some other subproject (or even the parent subproject), look for a wrap in the `subprojects/` dir in the source tree for that project
   - If there's no wrap, use the cached dependency
   - If there is a wrap and it's identical to the previous wrap (same git repo, same commit), use the cached dependency
   - If the wrap is not identical, error out and tell the user to use `meson wrap promote`
       * Optionally, we can print a list of all unique wraps that can be promoted

I think this will work very well in the ordinary case of all dependencies having their own wraps that point to `origin/master` of the same repository. Otherwise your first time building any such project will be several tautological `meson wrap promote` commands (same as now).

**Comment by nirbheek** [comment]
Fwiw, the behaviour I listed above would really make my life simpler while pitching Meson as a good build system to build-your-app-out-of-the-box-on-all-platforms. There is nothing else out there that can do this right now, and we're very close. :)

**Comment by jpakkane** [comment]
I'm not a fan of trying something risky and then aborting later on errors because it leaves stale data around. In this case step #2 _must_ write to the top level subproject directory (and not the subproject subproject dir). That means that either is must be cleaned out on errors or the configuration will succeed if it is run afresh because now the dependency is at the top subproject dir (which it should not be because there are two or more conflicting providers).

Having some sort of automatic promotion would be useful. We just have to think of a reliable and usable UX for it.

**Comment by nirbheek** [comment]
> I'm not a fan of trying something risky and then aborting later on errors because it leaves stale data around. In this case step #2 must write to the top level subproject directory (and not the subproject subproject dir). That means that either is must be cleaned out on errors or the configuration will succeed if it is run afresh because now the dependency is at the top subproject dir (which it should not be because there are two or more conflicting providers).

This indeed a problem, and I can think of a number of possible ways to overcome this:

1. Put information in the builddir that identifies the subproject dir as having been created for a sub-subproject wrap so we don't misidentify on reconfigure
    - This has the problem that new builddirs will misidentify it
1. Put that information in a hidden file inside the subproject dir that we write out
    - This will confuse the user if they do `rm -rf subprojects/foolib/* && tar -xf foolib.tar.xz -C subprojects/foolib/` or similar
1. Put that information in the `subprojects/` dir inside a file called `foolib.wrap-conferror` (or similar, we can bikeshed about the name)
1. Name the dir we place inside `subprojects/` something like `subprojects/meson-autowrap-foolib/` so that it is never autodetected

What do you think?

**Comment by jpakkane** [comment]
I don't like magic hidden state or things of that nature. How about starting simple. If a dependency is not found, Meson will search the tree for it and if it finds one or more it will print something like:

    Dependency [foo] was not available. However there are copies in the source tree.
    If you wish to use one of them, issue one of the following commands:

    meson wrap promote subprojects/foo/subprojects/bar
    meson wrap promote subprojects/baz/subprojects/bar

Then it will exit. People can then copypaste the command easily. This gives a fairly good UX and we can update it later once we get usage reports.

File: mesonbuild/interpreter.py [hunk 1]
```diff
@@ -2112,6 +2113,28 @@ def func_dependency(self, node, args, kwargs):
...
+        found = sprojs[dependency_name]
+        if len(found) > 1:
+            suffix = 'one of the following commands'
+        else:
+            suffix = 'the following command'
+        message = message_templ % (dependency_name, suffix)
+        cmds = []
+        command_templ = 'meson wrap promote '
+        for l in found:
+            cmds.append(command_templ + l[len(self.source_root)+1:])
```
Comment (ignatenkobrain) [inline mesonbuild/interpreter.py]: __[flake8]__
* [E226] missing whitespace around arithmetic operator

**Comment by nirbheek** [comment]
> I don't like magic hidden state or things of that nature. How about starting simple. If a dependency is not found, Meson will search the tree for it and if it finds one or more it will print something like:

If we're going to search through all the subprojects and find wraps, I think we should also parse them and if they are identical, we should promote one ourselves. That will address all issues I think. Could even add a comment at the end of the wrap file saying that meson automatically promoted it from the path `subprojects/blah/`.

**Comment by jpakkane** [comment]
The current impl ignores wrap files and only looks at subdirectories. Should add wrap file promotion, too....

File: mesonbuild/mesonlib.py [hunk 2]
```diff
@@ -760,6 +760,40 @@ def windows_proof_rmtree(f):
     # Try one last time and throw if it fails.
     shutil.rmtree(f)
 
+def unholder_array(entries):
+    result = []
+    entries = flatten(entries)
```
Comment (ignatenkobrain) [inline mesonbuild/mesonlib.py]: __[flake8]__
* [F821] undefined name 'flatten'

**Comment by NickeZ** [comment]
Would it be completely infeasable to mangle the symbols in a subproject so that multiple versions of the subproject could be used instead? The approach in this PR doesn't seem to scale well to larger projects and it seems to be in conflict with what rust does.

File: mesonbuild/interpreter.py [hunk 3]
```diff
@@ -2209,6 +2210,28 @@ def func_dependency(self, node, args, kwargs):
...
```
Comment (ignatenkobrain) [inline mesonbuild/interpreter.py]: __[Flake8]__

missing whitespace around arithmetic operator

File: mesonbuild/mesonlib.py [hunk 4]
```diff
@@ -881,6 +881,40 @@ def windows_proof_rmtree(f):
...
```
Comment (ignatenkobrain) [inline mesonbuild/mesonlib.py]: __[Flake8]__

undefined name 'flatten'

**Comment by jpakkane** [comment]
> it seems to be in conflict with what rust does

What Rust does it exactly the same as what NPM does. Which is just plain wrong.

**Comment by NickeZ** [comment]
> What Rust does it exactly the same as what NPM does. Which is just plain wrong.

I now that it might not be practically feasible. But if it was possible to rename a library's all functions to libname_X_Y_Z_funA() and change all the call locations in the "dependee". Why would that be a bad idea?

**Comment by nirbheek** [comment]
> I now that it might not be practically feasible. But if it was possible to rename a library's all functions to libname_X_Y_Z_funA() and change all the call locations in the "dependee". Why would that be a bad idea?

When I first read this suggestion, I was horrified, mortified, and petrified. Then I started smiling because it reminded me of https://xkcd.com/1425/

To make such a change, you need a static analyzer that has full information of all the ways in which a library may export a symbol, and all the ways in which a dependency may access a symbol. This static analyzer must do all the work of not just a preprocessor and a compiler, but also a virtual machine because you can access symbols at runtime with `dlsym`/`LoadLibrary` and with function pointers, ffi, `longjmp`, etc etc.

It is impossible to get this right, and the work required to make it work for C projects that use symbols in non-trivial ways is probably a year or more, and that's only because Clang already exists (or you'd have to invent it first).

**Comment by nirbheek** [comment]
See, for example: http://coccinelle.lip6.fr/ which is a multi-year project to do exactly this.

**Comment by NickeZ** [comment]
Interesting, yeah I guess you have to take this into account already when you create your language if you want this supported..

**Comment by jpakkane** [comment]
Since that is probably not happening soon, is this ready to merge?

File: mesonbuild/mesonlib.py (hunk 4 above)
Comment (nirbheek) [reply mesonbuild/mesonlib.py]: This flake8 issue seems to be important.

Comment (jpakkane) [reply mesonbuild/mesonlib.py]: This seems to be dead code. Removed.

**Comment by tp-m** [comment]
The version number in the docs addition needs to be updated. Otherwise I think it'd be nice to get this in. It will be incredibly useful already. We can add more smarts later.

**Pull request merged**